File: api-flask/app/idpoor.py
"""Collect and parse idpoor cambodia data."""
import requests

BASE_API = 'https://mop.idpoor.gov.kh/api/analytics/public/units'


def get_idpoor_response():
    """Get IDPoor location data."""
    datalist = dict()
    default_params = 'region=ALL&isOnDemand=ALL&year=ALL'

    rounds_uri = '{0}/KH?{1}&round=ALL'.format(BASE_API, default_params)
    resp = requests.get(rounds_uri)
    resp.raise_for_status()

    rounds_data = resp.json()['rounds']

    rounds = list(map(lambda data: int(data['name']), rounds_data))
    rounds.sort(reverse=True)

    # loop over reversed round years
    for round in rounds:
        provinces_uri = '{0}/KH/children?round={1}&{2}'.format(BASE_API, round, default_params)
        resp = requests.get(provinces_uri)
        resp.raise_for_status()

        provinces_data = resp.json()
        # one each year get provinces and get all communes per province
        for province in provinces_data:
            province_id = province['uuid']
            districts_uri = '{0}/{1}/children?round={2}&{3}'.format(
                BASE_API, province_id, round, default_params
            )
            resp = requests.get(districts_uri)
            resp.raise_for_status()

            districts_data = resp.json()

            for district in districts_data:
                district_id = district['uuid']
                communes_uri = '{0}/{1}/children?round={2}&{3}'.format(
                    BASE_API, district_id, round, default_params
                )
                resp = requests.get(communes_uri)
                resp.raise_for_status()

                communes_data = resp.json()

                for commune in communes_data:
                    key = int(commune['code'])
                    if key not in datalist.keys():
                        datalist[key] = {'Adm3_NCDD': str(key), **commune}

    return {'DataList': list(datalist.values())}

File: scripts/idpoor.ts
import * as fs from 'fs';
import * as path from 'path';
import axios from 'axios';
import { get } from 'lodash';

const ROOT_DIR = path.dirname(__dirname);
const BASE_URI = 'https://mop.idpoor.gov.kh/api/analytics/public/units';
const PARAMS = 'region=ALL&isOnDemand=ALL&year=ALL';

const fetchIdPoor = async () => {
  const roundsUri = `${BASE_URI}/KH?${PARAMS}&round=ALL`;
  try {
    // eslint-disable-next-line fp/no-mutating-methods
    const rounds = await (await axios.get(roundsUri)).data.rounds
      .map((r: object) => Number(get(r, 'name')))
      .sort((n: number, nn: number) => n - nn)
      .reduce((acc: number[], n: number) => [n, ...acc], []);

    const all = rounds.map(async (round: number) => {
      const pUri = `${BASE_URI}/KH/children?${PARAMS}&round=${round}`;
      try {
        const provinces = await (await axios.get(pUri)).data;

        const allD = provinces.map(async (d: object) => {
          const dId = get(d, 'uuid');
          const dUri = `${BASE_URI}/${dId}/children?${PARAMS}&round=${round}`;
          try {
            const districts = await (await axios.get(dUri)).data;

            const allC = districts.map(async (c: object) => {
              const cId = get(c, 'uuid');
              const cUri = `${BASE_URI}/${cId}/children?${PARAMS}&round=${round}`;
              try {
                const communes = await (await axios.get(cUri)).data;

                const allStats = communes.map(async (commune: object) => {
                  const communeId = Number(get(commune, 'uuid')).toString();
                  const communeCode = Number(get(commune, 'code')).toString();
                  const sUri = `${BASE_URI}/${communeId}/statistics?${PARAMS}&round=${round}`;
                  const formattedStats = {
                    Adm3_NCDD: communeCode,
                    name: get(commune, 'name'),
                    nameKm: get(commune, 'nameKm'),
                    total: get(commune, 'rural'),
                  };

                  try {
                    const stats = await (await axios.get(sUri)).data;

                    return {
                      ...formattedStats,
                      level1: get(stats[0], 'households'),
                      level2: get(stats[1], 'households'),
                    };
                  } catch {
                    return formattedStats;
                  }
                });

                return Promise.all(allStats);
              } catch {
                return null;
              }
            });

            return Promise.all(allC);
          } catch {
            return null;
          }
        });

        return Promise.all(allD);
      } catch {
        return null;
      }
    });

    return Promise.all(all);
  } catch {
    throw new Error('Failed to get rounds');
  }
};

(async () => {
  const COUNTRY: string = process.env.REACT_APP_COUNTRY as string;

  if (COUNTRY === 'cambodia') {
    const data = await fetchIdPoor();
    const datalist = JSON.stringify({
      DataList: data.flat(3).filter(d => d !== null),
    });
    console.warn(datalist);

    const dataDir = path.join(ROOT_DIR, `public/data/${COUNTRY}`);
    const file = path.join(dataDir, 'idpoor.json');
    fs.writeFile(file, datalist, (fileWriteError: Error | null) => {
      if (fileWriteError) {
        throw fileWriteError;
      }
    });
  } else {
    console.warn(`ID Poor scrip not suppored in ${COUNTRY}`);
  }

  console.warn('- Done');
})();
